FitGoal/FitGoal/Singup&Singin/LoginViewController.swift

```diff
@@ -0,0 +1,228 @@
+        let title = "Login".formattedText(
+            font: "Roboto-Bold",
+            size: 17,
+            color: .white,
+            kern: 0
+        )
+        button.setAttributedTitle(title, for: .normal)
+        return button
+    }()
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+    
+        view.backgroundColor = #colorLiteral(red: 0.9647058824, green: 0.9647058824, blue: 0.9647058824, alpha: 1)
+        view.addSubview(scrollView)
+        let views = [
+            backgroundView,
+            avatarView,
+            mainLabel,
+            loginButton,
+            loginForm,
+            socialMediaView,
+            linkToSignUp,
+        ]
+        scrollView.addMultipleSubviews(views)
+        setConstraints()
+        
+        linkToSignUp.delegate = self
+        
+        let tap = UITapGestureRecognizer(
+            target: self,
+            action: #selector(presentViewControler(_:))
+        )
+        loginButton.addGestureRecognizer(tap)
+        
+        let dismissKeyBoardTap = UITapGestureRecognizer(
+            target: self,
+            action: #selector(dismissKeyboard(_:))
+        )
+        scrollView.addGestureRecognizer(dismissKeyBoardTap)
+    
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(keyboardWillShow(notification:)),
+            name: UIResponder.keyboardWillShowNotification,
+            object: nil
+        )
+        
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(keyboardWillHide(notification:)),
+            name: UIResponder.keyboardWillHideNotification,
+            object: nil
+        )
+    }
+    
+    override func viewDidLayoutSubviews() {
+        super.viewDidLayoutSubviews()
+        scrollView.contentSize = view.frame.size
+        
+        socialMediaView.frame = CGRect(
+            x: 0,
+            y: 2/3 * view.bounds.maxY,
+            width: view.bounds.width,
+            height: 1/3 * view.bounds.height
+        )
+        
+        avatarView.frame = CGRect(
+            x: view.bounds.midX - 42,
+            y: 130,
+            width: 104,
+            height: 104
+        )
+        
+        backgroundView.frame = CGRect(
+            x: 0,
+            y: 0,
+            width: view.bounds.width,
+            height: 1/3 * view.bounds.height
+        )
+        
+        loginButton.frame = CGRect(
+            x: 16,
+            y: view.bounds.midY + 72,
+            width: view.bounds.width - 32,
+            height: 52
+        )
+        
+        loginButton.layer.cornerRadius = loginButton.bounds.height/2
+    }
+    
+    @objc private func dismissKeyboard(_ sender: UITapGestureRecognizer) {
+        switch sender.state {
+        case .ended:
+            loginForm.endEditing(true)
+        default:
+            return
+        }
```

Comment:
When you are handling an `UITapGestureRecognizer` you don't need to consider the state. You have to do that for other types of recognisers that have different stages (like dragging), but not with tap.